File: uniffi_bindgen/src/bindings/gecko_cpp/templates/macros.cpp
{# /* Return type... */ #}
{%- macro decl_ret_type(func) -%}
  {%- match func.throws() -%}
    {%- when Some with (err) -%}
      {%- match func.return_type() -%}
        {%- when Some with (type_) -%}
          Result<{{ type_|type_cpp }}, {{ err }}>
        {%- when None -%}
          Result<Ok, {{ err }}>
      {%- endmatch -%}
    {%- when None -%}
      {%- match func.return_type() -%}
        {%- when Some with (type_) -%}
          {{ type_|type_cpp }}
        {%- when None -%}
          void
      {%- endmatch -%}
  {%- endmatch -%}
{%- endmacro -%}

{# /* Calls an FFI function. */ #}
{%- macro to_ffi_call(func) -%}
  {%- call to_ffi_call_head(func, "err", "loweredRetVal_") -%}
  {%- call _to_ffi_call_tail(func, "err", "loweredRetVal_") -%}
{%- endmacro -%}

{# /* Calls an FFI function with an initial argument. */ #}
{%- macro to_ffi_call_with_prefix(prefix, func) %}
  RustError err = {0, nullptr};
  {% match func.ffi_func().return_type() %}{% when Some with (type_) %}const {{ type_|type_ffi }} loweredRetVal_ ={% else %}{% endmatch %}{{ func.ffi_func().name() }}(
    {{ prefix }}
    {%- let args = func.arguments() -%}
    {%- if !args.is_empty() %},{% endif -%}
    {%- for arg in args %}
    {{ arg.type_()|lower_cpp(arg.name()) }}{%- if !loop.last %},{% endif -%}
    {%- endfor %}
    , &err
  );
  {%- call _to_ffi_call_tail(func, "err", "loweredRetVal_") -%}
{%- endmacro -%}

{# /* Calls an FFI function without handling errors or lifting the return
      value. Used in the implementation of `to_ffi_call`, and for
      constructors. */ #}
{%- macro to_ffi_call_head(func, error, result) %}
  RustError {{ error }} = {0, nullptr};
  {% match func.ffi_func().return_type() %}{% when Some with (type_) %}const {{ type_|type_ffi }} {{ result }} ={% else %}{% endmatch %}{{ func.ffi_func().name() }}(
    {%- let args = func.arguments() -%}
    {%- for arg in args %}
    {{ arg.type_()|lower_cpp(arg.name()) }}{%- if !loop.last %},{% endif -%}
    {%- endfor %}
    {% if !args.is_empty() %}, {% endif %}&{{ error }}
  );
{%- endmacro -%}

{# /* Handles errors and lifts the return value from an FFI function. */ #}
{%- macro _to_ffi_call_tail(func, err, result) %}
  {%- match func.throws() %}
  {%- when Some with (name) %}
  auto maybeError = {{ name|class_name_cpp }}::FromConsuming(err);
  if (maybeError.isSome()) {
    return Err(maybeError.extract());
  }
  {% else %}
  if ({{ err }}.mCode) {
    MOZ_ASSERT(false);
    return {% match func.return_type() %}{% when Some with (type_) %}{{ type_|dummy_ret_value_cpp }}{% else %}{% endmatch %};
  }
  {% endmatch %}
  {%- match func.return_type() %}
  {%- when Some with (type_) %}
  {{ type_|type_cpp }} retVal_;
  DebugOnly<bool> ok_ = {{ type_|lift_cpp(result, "retVal_") }};
  MOZ_ASSERT(ok_);
  return std::move(retVal_);
  {% else %}{%- endmatch %}
{%- endmacro -%}

File: uniffi_bindgen/src/templates/macros.rs
{#
// Template to receive calls into rust.
#}

{%- macro to_rs_call(func) -%}
{{ func.name() }}({% call _arg_list_rs_call(func) -%})
{%- endmacro -%}

{%- macro to_rs_call_with_argname(arg_name, func, obj) -%}
    {{ func.name() }}(
    {%- if obj.threadsafe() %}
        {# threadsafe objects assume `&self` #}
        &{{- arg_name -}}
    {%- else -%}
        {# non-threadsafe objects must acquire the mutex we wrapped them in, allowing a `&mut self` #}
        &mut *{{- arg_name }}.lock().unwrap()
    {%- endif -%}
    {% if func.arguments().len() > 0 %}, {% call _arg_list_rs_call(func) -%}{% endif -%}
)
{%- endmacro -%}

{%- macro _arg_list_rs_call(func) %}
    {%- for arg in func.arguments() %}
        {%- if arg.by_ref() %}&{% endif %}
        {{- arg.name()|lift_rs(arg.type_(), true) }}
        {%- if !loop.last %}, {% endif %}
    {%- endfor %}
{%- endmacro -%}

{#-
// Arglist as used in the _UniFFILib function declations.
// Note unfiltered name but type_ffi filters.
-#}
{%- macro arg_list_ffi_decl(func) %}
    {%- for arg in func.arguments() %}
        {{- arg.name() }}: {{ arg.type_()|type_ffi -}}{% if loop.last %}{% else %},{% endif %}
    {%- endfor %}
    {% if func.arguments().len() > 0 %},{% endif %} err: &mut uniffi::deps::ffi_support::ExternError,
{%- endmacro -%}

{%- macro arg_list_decl_with_prefix(prefix, meth) %}
    {{- prefix -}}
    {%- if meth.arguments().len() > 0 %}, {# whitespace #}
        {%- for arg in meth.arguments() %}
            {{- arg.name() }}: {{ arg.type_()|type_rs -}}{% if loop.last %}{% else %},{% endif %}
        {%- endfor %}
    {%- endif %}
{%- endmacro -%}

{% macro return_type_func(func) %}{% match func.ffi_func().return_type() %}{% when Some with (return_type) %}{{ return_type|type_ffi }}{%- else -%}(){%- endmatch -%}{%- endmacro -%}

{% macro ret(func) %}{% match func.return_type() %}{% when Some with (return_type) %}{{ "_retval"|lower_rs(return_type, true) }}{% else %}_retval{% endmatch %}{% endmacro %}

{% macro construct(obj, cons) %}
    {%- if !obj.threadsafe() %}std::sync::Mutex::new({% endif -%}
    {{- obj.name() }}::{% call to_rs_call(cons) -%}
    {%- if !obj.threadsafe() %}){% endif -%}
{% endmacro %}

{% macro to_rs_constructor_call(obj, cons) %}
{% match cons.throws() %}
{% when Some with (e) %}
    uniffi::deps::ffi_support::call_with_result(err, || -> Result<_, {{ e }}> {
        let _new = {% call construct(obj, cons) %}?;
        let _arc = std::sync::Arc::new(_new);
        Ok({{ "_arc"|lower_rs(obj.type_(), obj.threadsafe()) }})
    })
{% else %}
    uniffi::deps::ffi_support::call_with_output(err, || {
        let _new = {% call construct(obj, cons) %};
        let _arc = std::sync::Arc::new(_new);
        {{ "_arc"|lower_rs(obj.type_(), obj.threadsafe()) }}
    })
{% endmatch %}
{% endmacro %}

{% macro to_rs_method_call(obj, meth) -%}
{% let receiver = meth.first_argument().name().to_string() %}
{% let receiver_type =  meth.first_argument().type_() %}
{% match meth.throws() -%}
{% when Some with (e) -%}
    uniffi::deps::ffi_support::call_with_result(
        err,
        || -> Result<_, {{ e }}> {
            let _obj = {{ receiver|lift_rs(receiver_type, obj.threadsafe()) }};
            let _retval = {{ obj.name() }}::{%- call to_rs_call_with_argname("_obj", meth, obj) -%}?;
            Ok({% call ret(meth) %})
        },
    )
{% else -%}
    uniffi::deps::ffi_support::call_with_output(
        err,
        || {
            let _obj = {{ receiver|lift_rs(receiver_type, obj.threadsafe()) }};
            let _retval = {{ obj.name() }}::{%- call to_rs_call_with_argname("_obj", meth, obj) -%};
            {% call ret(meth) %}
        },
    )
{% endmatch -%}
{% endmacro -%}

{% macro to_rs_function_call(func) %}
{% match func.throws() %}
{% when Some with (e) %}
uniffi::deps::ffi_support::call_with_result(err, || -> Result<{% call return_type_func(func) %}, {{e}}> {
    let _retval = {% call to_rs_call(func) %}?;
    Ok({% call ret(func) %})
})
{% else %}
uniffi::deps::ffi_support::call_with_output(err, || {
    let _retval = {% call to_rs_call(func) %};
    {% call ret(func) %}
})
{% endmatch %}
{% endmacro %}
